ykcompile/src/lib.rs

```diff
@@ -908,113 +508,216 @@ impl<TT> TraceCompiler<TT> {
         self.caller_save_restore();
 
         if let Some(d) = dest {
-            self.store(d, Location::Register(RAX.code()));
+            let dest_loc = self.iplace_to_location(d);
+            self.store_raw(
+                &dest_loc,
+                &Location::Register(RAX.code()),
+                SIR.ty(&d.ty()).size(),
+            );
         }
 
         Ok(())
     }
 
-    fn c_binop(&mut self, binop: &BinOp, op1: &Operand, op2: &Operand) -> Location {
-        match binop {
+    /// Load an IPlace into the given register. Panic if it doesn't fit.
+    fn load_reg_iplace(&mut self, reg: u8, src_ip: &IPlace) -> Location {
+        let dest_loc = Location::Register(reg);
+        let src_loc = self.iplace_to_location(src_ip);
+        self.store_raw(&dest_loc, &src_loc, SIR.ty(&src_ip.ty()).size());
+        dest_loc
+    }
+
+    fn c_binop(&mut self, dest: &IPlace, op: BinOp, opnd1: &IPlace, opnd2: &IPlace, checked: bool) {
+        match op {
             BinOp::Eq | BinOp::Ne | BinOp::Lt | BinOp::Le | BinOp::Gt | BinOp::Ge => {
-                self.c_condition(binop, op1, op2)
+                return self.c_condition(dest, &op, opnd1, opnd2);
             }
+            _ => {}
+        }
+
+        // We do this in three stages.
+        // 1) Copy the first operand into the temp register.
+        self.load_reg_iplace(*TEMP_REG, opnd1);
+
+        // 2) Add the second operand.
+        // FIXME replace with macro
+        let size = SIR.ty(&opnd1.ty()).size();
+        match op {
+            BinOp::Add => self.c_binop_add(dest, size, opnd2),
+            BinOp::Sub => self.c_binop_sub(dest, size, opnd2),
             _ => todo!(),
         }
-    }
 
-    fn c_condition(&mut self, binop: &BinOp, op1: &Operand, op2: &Operand) -> Location {
-        let (src1, ty) = match op1 {
-            Operand::Place(p) => self.place_to_location(&p, false),
-            Operand::Constant(Constant::Int(_)) => todo!(),
-            Operand::Constant(Constant::Bool(_b)) => todo!(),
-            Operand::Constant(c) => todo!("{}", c),
-        };
-        let src2 = match op2 {
-            Operand::Place(p) => self.c_place(&p),
-            Operand::Constant(Constant::Int(ci)) => self.c_constint(&ci),
-            Operand::Constant(Constant::Bool(_b)) => todo!(),
-            Operand::Constant(c) => todo!("{}", c),
-        };
+        // FIXME. Check for overflow here. Examine either CF or OF depending on signedness.
 
-        match (&src1, &src2) {
-            (Location::Register(reg1), Location::Register(reg2)) => match ty.size() {
-                1 => {
+        // 3) Move the result to where it is supposed to live.
+        // If it is a checked operation, then we have to build a (value, overflow?) tuple.
```

Comment:
`overflow?` -> `overflows` maybe? [I wondered initially if the `?` meant "I don't know if this is right or not!]